models/ssh_key.go

```diff
@@ -1030,3 +1058,275 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+		// First of ensure that the RootPath is present, and if not make it with 0700 permissions
+		// This of course doesn't guarantee that this is the right directory for authorized_keys
+		// but at least if it's supposed to be this directory and it doesn't exist and we're the
+		// right user it will at least be created properly.
+		err := os.MkdirAll(setting.SSH.RootPath, 0700)
+		if err != nil {
+			log.Error("Unable to MkdirAll(%s): %v", setting.SSH.RootPath, err)
+			return err
+		}
+	}
+
+	fPath := filepath.Join(setting.SSH.RootPath, authorizedPrincipalsFile)
+	f, err := os.OpenFile(fPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Note: chmod is noop on Windows.
+	if !setting.IsWindows {
+		fi, err := f.Stat()
+		if err != nil {
+			return err
+		}
+
+		// .ssh directory should have mode 700, and authorized_keys file should have mode 600.
+		if fi.Mode().Perm() > 0600 {
+			log.Error(authorizedPrincipalsFile+"file has unusual permission flags: %s - setting to -rw-------", fi.Mode().Perm().String())
+			if err = f.Chmod(0600); err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, key := range keys {
+		if _, err = f.WriteString(key.AuthorizedString()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetPrincipalKeyByID returns public key by given ID.
+func GetPrincipalKeyByID(keyID int64) (*PublicKey, error) {
+	return GetPublicKeyByID(keyID)
+}
+
+// CheckPrincipalKeyString checks if the given public key string is recognized by SSH.
+// It returns the actual public key line on success.
+func CheckPrincipalKeyString(content string) (_ string, err error) {
+	if setting.SSH.Disabled {
+		return "", ErrSSHDisabled{}
+	}
+
+	content = strings.TrimRight(content, "\n\r")
+	if strings.ContainsAny(content, "\n\r") {
+		return "", errors.New("only a single line with a single principal please")
+	}
+
+	// remove any unnecessary whitespace now
+	content = strings.TrimSpace(content)
+
+	return content, nil
+}
+
+// deletePrincipalKeys does the actual key deletion but does not update authorized_principals file.
+func deletePrincipalKeys(e Engine, keyIDs ...int64) error {
+	if len(keyIDs) == 0 {
+		return nil
+	}
+
+	_, err := e.In("id", keyIDs).Delete(new(PublicKey))
+	return err
+}
+
+// DeletePrincipalKey deletes SSH key information both in database and authorized_principals file.
+func DeletePrincipalKey(doer *User, id int64) (err error) {
+	key, err := GetPrincipalKeyByID(id)
+	if err != nil {
+		return err
+	}
+
+	// Check if user has access to delete this key.
+	if !doer.IsAdmin && doer.ID != key.OwnerID {
+		return ErrKeyAccessDenied{doer.ID, key.ID, "public"}
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+	if err = sess.Begin(); err != nil {
+		return err
+	}
+
+	if err = deletePrincipalKeys(sess, id); err != nil {
+		return err
+	}
+
+	if err = sess.Commit(); err != nil {
+		return err
+	}
+	sess.Close()
+
+	return RewriteAllPrincipalKeys()
+}
+
```

Comment:
You don't need these functions if you're not going to enforce that the key is a particular type.

Instead at the end of `DeletePublicKey` just check the Type of the key you obtained and do a rewrite:

```go
	if key.Type == KeyTypePrincipal {
		return RewriteAllPrincipalKeys()
	}
```

Reply:
Yes, I understand I could do that, but this makes it harder to read where those functions are used, having it differ for principal and publickey makes it more readable imo.

But if you want it removed, I'll do that.